### DISCUSSION TITLE
Add dart lsp for flutter developer

### DISCUSSION THREAD

**Discussion by n2ng-dev** [ayamir/nvimdots#539] [Q&A] [answered]
Hii I'm beginner Flutter Dev

How to add dartlsp and this plugins in this NVIM setup ?

plugins : [flutter-tools.nvim](https://github.com/akinsho/flutter-tools.nvim)

thanks

**Comment by aarnphm** [comment]
YOu can add it locally similar to rust-tools.nvim.

**Reply by n2ng-dev** [reply]
Hai @aarnphm can you help me explaind it ?

**Reply by aarnphm** [reply]
See https://github.com/ayamir/nvimdots/wiki for introduction. We recently introduce dotstutor to help with using the setup.

This is where we add support for rust-tools https://github.com/ayamir/nvimdots/blob/b08a174aac580f19a941899dd78152f05fd96a6d/lua/modules/plugins/lang.lua#L9. You can do the same for flutter-tools.nvim

**Reply by n2ng-dev** [reply]
I'll try,
but I add new file and folder its not show in File Explorer

**Reply by ayamir** [reply]
It's because we set it in `.gitignore`, you can press `I` to see hidden folders in `nvim-tree`.

**Reply by n2ng-dev** [reply]
Hai @ayamir @aarnphm 
thanks for replay

I've followed the tutorial on adding plugins on the wiki
but there is an error when open & save dart file

how to makesure plugins load & run ?

--setup plugin 
![Screenshot 2023-02-24 at 19 47 53](https://user-images.githubusercontent.com/12078569/221182765-8b1631dc-f06f-4988-b3a6-1fda89a644fd.png)

--add plugin
![Screenshot 2023-02-24 at 19 47 53](https://user-images.githubusercontent.com/12078569/221183025-c97d7b8c-7b8f-4314-9f5c-6b17fbeefa4d.png)

--error save file
![Screenshot 2023-02-24 at 17 52 30](https://user-images.githubusercontent.com/12078569/221183147-f9ff0634-fbad-440b-b09b-a4b200cb56d6.png)

--error open project
![Screenshot 2023-02-24 at 17 58 44](https://user-images.githubusercontent.com/12078569/221183254-66e9337a-3742-4389-aea1-cec26d516f01.png)

--LspInfo
![Screenshot 2023-02-24 at 18 00 54](https://user-images.githubusercontent.com/12078569/221183299-fedbefc1-d967-48f3-8ac5-9577786e9b3b.png)


many thanks for your time to help me

**Reply by ayamir** [reply]
`:MasonInstall dart` to install `dart` as lsp.

**Reply by n2ng-dev** [reply]
@ayamir `:MasonInstall dart` error  "dart is not a valid package"

![image](https://user-images.githubusercontent.com/12078569/221340505-9d2de567-e2f0-4d31-8189-103544b84b0b.png)

**Reply by ayamir** [reply]
Sorry, here is a answer from author of `mason.nvim`: https://github.com/williamboman/mason.nvim/issues/136#issuecomment-1193913116

**Comment by ayamir** [comment, marked as answer]
@Nanank-kho I have implemented it at https://github.com/ayamir/nvimdots/commit/a087e1d5dba57ce492ee83bf9849e7fc151c9ba9 as a sample config for lsps which are not supported by mason.nvim but supported by nvim-lspconfig. Just git pull and everything will be fine.

**Reply by n2ng-dev** [reply]
git pull done, but still got error 

-- open dart file
![Screenshot 2023-02-25 at 17 32 09](https://user-images.githubusercontent.com/12078569/221352878-26d13c29-5223-4e25-b2e1-09551ead6d11.png)

--save dart file
![Screenshot 2023-02-25 at 17 28 05](https://user-images.githubusercontent.com/12078569/221352900-b542e633-4e4a-4036-b76e-d2c544a8cb84.png)

**Reply by aarnphm** [reply]
Make sure dart is available in PATH. See https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/dartls.lua

**Reply by n2ng-dev** [reply]
yeah this error because open file dart in file explorer with neovide,

in case open folder project flutter with `nvim` all works fine

how to open project folder with neovide like nvim ?

many thanks...

**Reply by ayamir** [reply]
try launch `neovide` from your shell.

**Reply by n2ng-dev** [reply]
I have try, but error command not found

![image](https://user-images.githubusercontent.com/12078569/221364232-a0c4c712-228c-4714-b85f-c1f0bd04d648.png)

**Reply by n2ng-dev** [reply]
ohhh its clear, I remove app and install again with hombrew

**Reply by n2ng-dev** [reply]
now when save file show two notification and one error _LSP null-ls timeout_
![Screenshot 2023-02-25 at 22 22 23](https://user-images.githubusercontent.com/12078569/221365304-fdafa41e-2021-44e3-8e8c-0dbe035990eb.png)

**Reply by charliie-dev** [reply]
if u want to use dart lsp's formatting ability, u can disable null-ls' related server in null_la.lua. but keep in mind we have a drafting pr that might change the behavior. So keep up-to-date with our commits!

**Reply by charliie-dev** [reply]
@Nanank-kho The pr had been merged. Please set `dart_format` in this table: https://github.com/ayamir/nvimdots/blob/54e85577ab5252bf742c8497989cea409fe41b53/lua/core/settings.lua#L74-L83

If u have dart runtime installed, then adding this and https://github.com/ayamir/nvimdots/blob/54e85577ab5252bf742c8497989cea409fe41b53/lua/modules/configs/completion/lsp.lua#L103-L108 should do the work.

**Comment by aarnphm** [comment]
you probably can start the lsp with `vim.lsp.start`. I don't think we are using languageclient anymore. nvimdots will use the builtin LSP for the foreseeable future.

See https://neovim.io/doc/user/lsp.html fyi

**Comment by ayamir** [comment]
STOP USE CHATGPT TO GENERATE HELPLESS ANSWER.
